dbt-redshift/src/dbt/adapters/redshift/connections.py

```diff
@@ -506,46 +515,101 @@ def execute(
         fetch: bool = False,
         limit: Optional[int] = None,
     ) -> Tuple[AdapterResponse, "agate.Table"]:
-        sql = self._add_query_comment(sql)
-        _, cursor = self.add_query(sql, auto_begin)
-        response = self.get_response(cursor)
-        if fetch:
-            table = self.get_result_from_cursor(cursor, limit)
-        else:
-            from dbt_common.clients import agate_helper
+        def _handle_execute_exception(
+            e: Exception, retries: int, backoff: int, retry_all: bool
+        ) -> Tuple[int, int]:
+            oid_not_found_msg = "could not open relation with OID"
+            if retries == 0:
+                raise e
+            if oid_not_found_msg in str(e):
+                pass
+            elif isinstance(e, redshift_connector.Error) and retry_all:
+                pass
+            else:
+                logger.debug(f"Not retrying error: {e}")
+                raise e
+            logger.debug(f"Retrying query due to error: {e}")
+            retries -= 1
+            # we need to actually close and open to get a new connection
+            # otherwise no queries will succeed on this connection
+            self.close(self.get_thread_connection())
+            self.open(self.get_thread_connection())
+            time.sleep(backoff)
+            # return with exponential backoff
+            return retries, min(backoff * 2, 60)
+
+        def _execute_internal(
```

Comment:
I think I might prefer this as _try_execute_loop but maybe that's just too rusty.